### ISSUE TITLE
Implement changed error hierarchy from #4544

### ISSUE THREAD

**Issue by danielmewes** [rethinkdb/rethinkdb#4559]
See https://github.com/rethinkdb/rethinkdb/issues/4544#issuecomment-123514164

**Comment by mlucy** [comment]
@danielmewes -- this isn't assigned to anyone.  Are we going to have the person in charge of each driver do this, or should one person just sit down and do all of them do you think?

**Comment by danielmewes** [comment]
@mlucy I don't really care, though I think it will be easier if one person updates all the drivers at once. 
Since all of your 2.1 issues are now in code review or waiting for test fixes (or for this), would you mind picking this one?

**Comment by larkost** [comment]
I just checked in some changes to the testing system to make the polyglot test system actually work for this type of test, and it looks like there is some issue in how the JavaScript driver is picking up error types. For instance this is coming up in the `polyglot/changefeeds/edge.js` a few times where one of the new subclasses is not being recorded but rather `RqlRuntimeError` being returned.

**Comment by mlucy** [comment]
@larkost -- I broke that out into a separate issue so it won't get lost.

**Comment by deontologician** [comment]
Should the TimeoutError be in new drivers? The comment next to it seems to indicate it should only be there for backwards compatibility

**Comment by danielmewes** [comment]
@deontologician I think it's what the Python driver throws if waiting on a cursor times out (http://rethinkdb.com/api/python/next/). If there's a native error type for timeouts in the language that can be used instead. Are you asking because of the Java driver?

**Comment by deontologician** [comment]
yeah, I added some timeout code yesterday, and didn't create a ReqlTimeoutError based on the note next to it, I just wanted to clarify that was the intention

**Comment by deontologician** [comment]
OK, new query: what happens to the CLIENT_ERROR responsetype? Is it renamed something else? The previous thread didn't mention it

**Comment by danielmewes** [comment]
Good question. It should probably just be another error type `ReqlClientError` under `ReqlError`.

**Comment by mlucy** [comment]
I made the appropriate changes to the protobuf file, the Ruby driver and the tests in the `mlucy_4559` branch.  @deontologician, @grey -- could I talk you into doing the changes for the JS driver and the Python driver respectively?  I can do it if you guys don't have time before the release, but there were a few small subtleties in the Ruby driver changes (where to detect `AuthError`, how to do the aliasing for backward-compatibility) that I don't quite trust myself to do properly in the drivers I don't understand as well.

**Comment by mlucy** [comment]
Er, I mean @Tryneus .

**Comment by Tryneus** [comment]
I can probably do the Python driver.

**Comment by deontologician** [comment]
No problem

**Comment by VeXocide** [comment]
I did the initial Python changes, and would be happy to pick that up as well.

**Comment by VeXocide** [comment]
Just discussed this with @danielmewes, I'll take care of the Python changes

**Comment by deontologician** [comment]
What branch are the ruby changes on?

**Comment by deontologician** [comment]
Sorry, you mentioned the branch above, nevermind!

**Comment by deontologician** [comment]
OK, changes are up on josh/4559 I merged in next since @mlucy 's branch didn't have the fix for #4631 in it. I made a couple of assumptions in places where things weren't completely clear:

- I added the ReqlClientError as a direct subclass of ReqlError as @danielmewes suggested
- The js driver had a superclass called "RqlServerError" that isn't in the new hierarchy, I made it an alias for ReqlRuntimeError since it's all the errors with a backtrace from the server
- @mlucy made Rql -> Reql aliases for all of the error classes, but many of the exceptions only ever existed before on the 2.1 branch, so there's no reason to make them backward compatible in that way. So I only created aliases for `DriverError`, `ServerError`, `RuntimeError`, `CompileError` and `ClientError` since they existed pre-2.1

**Comment by neumino** [comment]
In https://github.com/rethinkdb/rethinkdb/issues/4544#issuecomment-123514164, some errors (like ReqlAvailabilityError) don't exist in the ql2.proto file.
Are the drivers supposed to parse the error for those types?

**Comment by deontologician** [comment]
Reql availability doesn't show up by itself, its just a superclass for the two op errors, in case you want to catch either one

**Comment by neumino** [comment]
Oh, thanks that make sense :)

**Comment by VeXocide** [comment]
Merged the Python changes into `next` via commit 10f15ddddb475560991d5fdc1a6b5d8b13d289bb.

**Comment by deontologician** [comment]
ok, should be good to go now
